src/core/stream.c

```diff
@@ -593,7 +607,9 @@ QuicStreamTryCompleteShutdown(
         //
         // Indicate the stream is completely shut down to the connection.
         //
-        QuicStreamSetReleaseStream(&Stream->Connection->Streams, Stream);
+        if (!Stream->Flags.DelayFCUpdate || Stream->Flags.HandleClosed) {
```

Comment:
If HandleClose == true, this function, QuicStreamTryCompleteShutdown, is called by QuicStreamClose, right? QuicStreamClose guarantees that QuicStreamSetReleaseStream will be called if DelayFCUpdate is set. Wouldn't we end up calling QuicStreamSetReleaseStream twice?

Reply:
This case is tricky. If an app shuts down (abortively) a stream using the `QUIC_STREAM_SHUTDOWN_FLAG_IMMEDIATE` flag then we will immediately give them the shutdown complete event and they can close the handle, but MsQuic will continue to track the stream until we've delivered (and received the ACK for) the corresponding stream frames over the wire. In this case, the `QuicStreamClose` call will run before this code path, but we don't want to release the stream there. Later, when everything is complete, this code path will be run finally, but since the handle has already been closed, we need to release the stream now.

Reply:
I see.